pyrevolve/pyrevolve.py

```diff
@@ -0,0 +1,191 @@
+        for i in range(self.__n_snapshots):
+            try:
+                allData = self.__container[i].data
+                for var in allData:
+                    print(allData[var].data),
+                    print(self.__head_fwd.data[var].data)
+            except:
+                pass
+
+    @property
+    def n_snapshots(self):
+        return self.__n_snapshots
+
+
+class Revolver(object):
+    """
+    This should be the only user-facing class in here. It manages the
+    interaction between the operators passed by the user, and the data storage.
+
+    What needs to be checkpointed? At the moment, pyrevolve requests from the
+    forward and reverse operators a list of their inputs and outputs. The
+    variables that are checkpointed are then given by the intersection of
+    forward_operator.input and forward_operator.output.
+    The rationale is that the forward computation needs to have all its input
+    variables to resume the computation, but of those, only the ones that are
+    modified over time need to be checkpointed.
+
+    Todo:
+        * Reverse operator is always called for a single step. Change this.
+        * Avoid redundant data stores if higher-order stencils save multiple
+          time steps, and checkpoints are close together.
+        * Only offline single-stage is supported at the moment.
+        * Give users a good handle on verbosity.
+
+    Nice to have, but too far away for a Todo:
+        In Algorithmic Differentiation, TBR (to-be-recorded) is a better way
+        to decide which variables need to be stored. Only variables that have
+        a nonlinear influence on the result are important, and only those parts
+        of the forward operator that compute such variables need to be
+        repeated. This would require a full analysis of the forward operator,
+        and the creation of a simplified forward operator for checkpointing,
+        that has all computations with no effect on the adjoint gradients
+        removed.
+    """
+
+    def __init__(self, fwd_operator, rev_operator, timesteps=None):
+        """Initialise checkpointer for a given forward- and reverse operator, a
+        given number of time steps, and a given storage strategy. The number of
+        time steps must currently be provided explicitly, and the storage must
+        be the single-staged memory storage."""
+        if(timesteps is None):
+            raise Exception("Online checkpointing not yet supported. Specify \
+                              number of time steps!")
+        self.storage = MemoryStorage(cr.adjust(timesteps))
+        storage_disk = None  # this is not yet supported
+        # We use the crevolve wrapper around the C++ Revolve library.
+        self.ckp = cr.CRevolve(self.storage.n_snapshots,
+                               timesteps, storage_disk)
+
+        self.fwd_operator = fwd_operator
+        self.rev_operator = rev_operator
+        fwd_in = fwd_operator.input_params
```

Comment:
We should probably throw a nice error message here if these arguments are not present on the provided operators. This might be left for a later PR though.